Server/Game/tests.py

```diff
@@ -1,12 +1,162 @@
+			invalid_perk_list += "fake_perk,"
+			invalid_perk_str += '"fake_perk",'
+		invalid_perk_list = invalid_perk_list.strip(",")
+		invalid_perk_str = invalid_perk_str.strip(",")
+		valid_perk_list = '"Extra Money", "Forest Fighter", "Mountain Fighter"'
+		too_many_perks = valid_perk_list + ', "Strong Arrows"'
+
+		# Test no units
+		self.channel.send('{"Command":"ST","Leader":"Sniper","Ability":"Extra Range","Perks":[]}')
+		result = self.channel.receive()
+		self.assertEqual(result, json.dumps({"Success":False,"Error":"The unit information is incomplete."}))
+
+		# Test no perks
+		self.channel.send('{"Command":"ST","Units":[' + valid_unit_list + '],"Leader":"Sniper","Ability":"Extra Range"}')
+		result = self.channel.receive()
+		self.assertEqual(result, json.dumps({"Success":False,"Error":"The perk information is incomplete."}))
+
+		# Test too few units selected
+		self.channel.send('{"Command":"ST","Units":["Archer"],"Leader":"Sniper","Ability":"Extra Range","Perks":[]}')
+		result = self.channel.receive()
+		self.assertEqual(result, json.dumps({"Success":False,"Error":"You must select "
+			+ str(version.unit_count) + " units, only 1 chosen."}))
+
+		# Test too many units selected
+		self.channel.send('{"Command":"ST","Units":[' + too_many_units + '],"Leader":"Sniper","Ability":"Extra Range","Perks":[]}')
+		result = self.channel.receive()
+		self.assertEqual(result, json.dumps({"Success":False,"Error":"Too many units have been selected (9)."}))
+
+		# Test missing leader
+		self.channel.send('{"Command":"ST","Units":[' + valid_unit_list + '],"Ability":"Extra Range","Perks":[]}')
+		result = self.channel.receive()
+		self.assertEqual(result, json.dumps({"Success":False,"Error":"The leader information is incomplete."}))
+
+		# Test missing ability
+		self.channel.send('{"Command":"ST","Units":[' + valid_unit_list + '],"Leader":"Sniper","Perks":[]}')
+		result = self.channel.receive()
+		self.assertEqual(result, json.dumps({"Success":False,"Error":"The leader information is incomplete."}))
+
+		# Test invalid leader
+		self.channel.send('{"Command":"ST","Units":[' + valid_unit_list + '],"Leader":"fake_leader","Ability":"Extra Range","Perks":[]}')
+		result = self.channel.receive()
+		self.assertEqual(result, json.dumps({"Success":False,"Error":"The leader information provided is invalid."}))
+
+		# Test invalid ability
+		self.channel.send('{"Command":"ST","Units":[' + valid_unit_list + '],"Leader":"Sniper","Ability":"fake_ability","Perks":[]}')
+		result = self.channel.receive()
+		self.assertEqual(result, json.dumps({"Success":False,"Error":"The leader information provided is invalid."}))
+
+		# Test invalid ability-leader pair
+		self.channel.send('{"Command":"ST","Units":[' + valid_unit_list + '],"Leader":"Sniper","Ability":"Steal","Perks":[]}')
+		result = self.channel.receive()
+		self.assertEqual(result, json.dumps({"Success":False,"Error":"The Sniper cannot use the ability Steal."}))
+
+		endTestLog("test1_set_team_bad_json_input")
```

Comment:
Can we remove this endTestLog function, looks like maybe it was a copy-paste mistake seeing as there are more statements after this line.